src/linear_algebra/Hermitian_inner_product.lean

```diff
@@ -0,0 +1,669 @@
+    sqr_sqrt (inprod_self_re_nonneg y), inprod_self_add],
+  unfold herm_ortho at H,
+  rw [←conj_inprod x y, H, conj_zero,
+    field.add_zero, field.add_zero, add_re],
+end
+
+def is_ortho_set (s : set α) :=
+∀ x y ∈ s, (x ⊥ y) ↔ x ≠ y 
+
+def is_orthonormal_set (s : set α) :=
+is_ortho_set s ∧ ∀ x ∈ s, is_normalised x
+
+noncomputable def proj (x y : α) :=
+((x ₀ y)/( ↑|y|*|y| )) • y
+
+lemma zero_proj (x : α) :
+proj 0 x = 0 := by dunfold proj; simp
+
+lemma proj_zero (x : α) :
+proj x 0 = 0 := by dunfold proj; simp
+
+lemma proj_self_eq_self (x : α) :
+proj x x = x :=
+begin
+  have ho : x = 0 ∨ x ≠ 0,
+  { apply em },
+  dunfold proj,
+  cases ho,
+  { rw ho,
+    simp },
+
+  { rw [←of_real_mul, of_real_herm_norm_mul_self,
+      div_self (inprod_self_ne_zero_iff.mpr ho), one_smul] }
+end
+
+lemma smul_proj (x y : α) {a : ℂ} : 
+proj (a • x) y = a • (proj x y) :=
+begin
+  dunfold proj,
+  rw [inprod_smul_left, smul_smul],
+  ring,
+end
+
+lemma proj_smul (x y : α) {a : ℂ} (ha : a ≠ 0) :
+proj x (a • y) = proj x y := 
+begin
+  have Hy : y = 0 ∨ y ≠ 0,
+  { apply em },
+  cases Hy,
+  { rw [Hy, smul_zero] },
+
+  { dunfold proj,
+    rw [inprod_smul_right, herm_norm_smul, of_real_mul, smul_smul],
+    suffices :  ((↑(abs a) * ↑|y| * (↑(abs a) * ↑|y| ))⁻¹ * x ₀ y * conj a * a) • y = ((↑|y| * ↑|y| )⁻¹ * x ₀ y) • y,
+    { rw [div_eq_inv_mul, div_eq_inv_mul,
+        field.mul_comm (conj a), ←field.mul_assoc],
+      exact this },
+    { rw [field.mul_assoc, ←field.mul_comm a, mul_conj, 
+        ←field.mul_assoc ( ↑(abs a) * ↑|y| ), field.mul_comm (↑(abs a)),
+        field.mul_assoc ( ↑|y| ), ←of_real_mul, mul_self_abs,
+        field.mul_comm ( ↑|y| ), field.mul_comm, field.mul_assoc,
+        mul_inv_eq, field.mul_comm, field.mul_assoc ((↑|y| * ↑|y| )⁻¹),
+        field.mul_comm (↑(norm_sq a))⁻¹, field.mul_assoc, field.mul_assoc (x ₀ y),
+        field.inv_mul_cancel, field.mul_one],
+      refl,
+      
+      exact of_real_ne_zero.mpr ((iff_false_right ha).mp (norm_sq_eq_zero)),
+      exact (mul_ne_zero (of_real_ne_zero.mpr ((norm_ne_zero_iff_ne_zero).mpr Hy)) (of_real_ne_zero.mpr ((norm_ne_zero_iff_ne_zero).mpr Hy))),
+      exact of_real_ne_zero.mpr ((iff_false_right ha).mp (norm_sq_eq_zero)) }}
+end
+
+lemma norm_proj_le (x y : α) : 
+|proj x y| ≤ |x| :=
+begin
+  have Hy : y = 0 ∨ y ≠ 0,
+  { apply em },
+  cases Hy,
+  { rw [Hy, proj_zero, herm_norm_zero],
+    exact @norm_nonneg _ _ x },
+
+  { dunfold proj,
+    rw [herm_norm_smul, complex.abs_div, ←of_real_mul,
+      abs_of_real, abs_mul_self, div_mul_comm, division_def,
+      mul_inv', ←field.mul_assoc, field.mul_inv_cancel,
+      field.one_mul, field.mul_comm, ←division_def, div_le_iff],
+     
+      exact abs_inprod_le_mul_herm_norm x y,
+      exact (norm_pos_iff y).mpr Hy,
+      exact norm_ne_zero_iff_ne_zero.mpr Hy }
+end
+
+lemma ortho_imp_proj_eq_zero {x y : α} :
+(x ⊥ y) → proj x y = 0 := 
+begin
```

Comment:
You could move `(H : x ⊥ y)` before the `:` into the assumption list.
The proof would become something like
```
delta herm_ortho at H,
delta proj,
simp [H]
```